apiserver/params/controller.go

```diff
@@ -132,10 +136,138 @@ type ControllerVersionResults struct {
 	GitCommit string `json:"git-commit"`
 }
 
-// DashboardInfo holds the results from an api call
-// to get address info for the juju dashboard.
-type DashboardInfo struct {
-	Addresses []string `json:"addresses"`
-	UseTunnel bool     `json:"use-tunnel"`
-	Error     *Error   `json:"error,omitempty"`
+const (
+	// DashboardConnectionTypeProxy is the type key used for proxy connections
+	DashboardConnectionTypeProxy = "proxy"
+
+	// DashboardConnectionTypeSSHTunnel is the type key used for ssh connections
+	DashboardConnectionTypeSSHTunnel = "ssh-tunnel"
+)
+
+// DashboardConnectionProxy represents a proxy connection to the Juju Dashboard
+type DashboardConnectionProxy struct {
+	Proxier       proxy.Proxier   `json:"proxier"`
+	proxierConfig json.RawMessage `json:"-"`
+	ProxierType   string          `json:"type"`
+}
+
+// ProxierFactory is an interface type representing a factory that can make a
+// new juju proxier from the supplied type and JSON config.
+type ProxierFactory interface {
+	ProxierFromJSONDataBag(proxierType string, config json.RawMessage) (proxy.Proxier, error)
+}
+
+// DashboardConnectionSSHTunnel represents an ssh tunnel connection to the Juju
+// Dashboard
+type DashboardConnectionSSHTunnel struct {
+	Host string `json:"host"`
+	Port string `json:"port"`
+}
+
+// DashboardConnection interface represents a generic interface for establishing
+// dashboard connections.
+type DashboardConnection interface {
+	Type() string
+}
+
+// DashboardConnectionInfo holds the information necassery
+type DashboardConnectionInfo struct {
+	Connection     DashboardConnection `json:"connection"`
+	ConnectionType string              `json:"connection-type"`
+	Error          *Error              `json:"error,omitempty"`
+}
+
+// NewDashboardConnectionProxy constructs a new DashboardConnectionProxy from
+// the supplied proxier.
+func NewDashboardConnectionProxy(proxier proxy.Proxier) *DashboardConnectionProxy {
+	proxierType := ""
+	if proxier != nil {
+		proxierType = proxier.Type()
+	}
+	return &DashboardConnectionProxy{
+		Proxier:     proxier,
+		ProxierType: proxierType,
+	}
+}
+
+// Type implements the DashboardConnection interface
+func (_ *DashboardConnectionProxy) Type() string {
+	return DashboardConnectionTypeProxy
+}
+
+// Type implements the DashboardConnection interface
+func (_ *DashboardConnectionSSHTunnel) Type() string {
+	return DashboardConnectionTypeSSHTunnel
+}
+
+// UnmarshalJSON implements the encoding/json Unmarshaller interface
+func (d *DashboardConnectionInfo) UnmarshalJSON(data []byte) error {
+	wireFormat := &struct {
+		Connection     json.RawMessage `json:"connection"`
```

Comment:
Do we do this anywhere else inside juju facades? This circumvents the schema generation, so I'd prefer just having two pointer variables.